### DISCUSSION TITLE
Constrain variable types in pattern matching

### DISCUSSION THREAD

**Discussion by dubiousconst282** [dotnet/csharplang#6211] [Language Ideas] [answered]
Currently, pattern matching a variable against a type incurs it to be redeclared into a new variable of the desired type, which is a bit of a chore:

```cs
Instruction inst = ...
if (inst is StoreVarInst store) {
    Foo(store.Dest);
}
//...
var result = inst switch {
    BinaryInst bin  => FoldBinary(bin.Op, bin.Left, bin.Right),
    CompareInst cmp => FoldCompare(cmp.Op, cmp.Left, cmp.Right),
    //...
    _ => null
};
```

It would be convenient if the original variable was "constrained" into the type being matched, so it would behave as a normal variable of that type inside the scope:

```cs
if (inst is StoreVarInst) {
    //`inst` is constrained to `StoreVarInst` inside this scope, so accessing `Dest` is ok
    Foo(inst.Dest);
}
//...
var result = inst switch {
    BinaryInst => FoldBinary(inst.Op, inst.Left, inst.Right),
    //...
};
```

Reassignments would be restricted to the type being matched, but I don't see a problem with them being disallowed.

**Comment by HaloFour** [comment, marked as answer]
> ```cs
>  if (inst is StoreVarInst) {
>     //`inst` is constrained to `StoreVarInst` inside this scope, so accessing `Dest` is ok
>     Foo(inst.Dest);
> }
> ```

This is already legal C# code and to reinterpret the type of the variable would be a breaking change:

[SharpLab](https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAICYCMBYAUHgQQAIBDIgXiKgFMB3IgIQAoBKAbjzwEsAzIpslwDOjIgCMWRAN54ic8QDoAsqzZEA9OqIAHMFyjARDPAF9ufAUWGNJM3PNLLVGrbv2GiBU51wZP02XlUAGYiVAAWIhVJcgA+MMwATiYAIgIU9m98X3RREH87BxCqOjDI6Ip41CTUhgyOXBMgA=)

**Reply by dubiousconst282** [reply]
Ah! I just realized that `is` came before redeclaration patterns, and it's probably why we have it this way.

Looking at your example, I think it would break in even simpler cases like method overloading.
